File: hdf-converters/converters/oscal-to-hdf/go/converter_profile.go
package oscal

import (
	"fmt"
	"strings"

	shared "github.com/mitre/hdf-libs/hdf-converters/shared/go"
	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go"
)

// ConvertProfileToHDF resolves an OSCAL Profile against a provided catalog and
// converts the result to an HDFBaseline. The profile's import directives select
// which controls from the catalog to include.
//
// This handles:
//   - A single catalog import
//   - include-controls with with-ids filtering
//   - alter directives (add/remove parts and props on controls)
//   - set-parameters from modify section
//   - merge as-is (preserve catalog structure)
//
// It does NOT handle:
//   - Multiple imports (returns an error with guidance)
//   - Nested profile imports (profile importing another profile)
//   - Complex merge strategies
//
// For full profile resolution, use NIST's oscal-cli or pre-resolved catalogs.
func ConvertProfileToHDF(profileInput, catalogInput []byte, converterVersion string) (*hdf.HDFBaseline, error) {
	profileDoc, err := ParseOscalDocument(profileInput, "profile", "oscal-profile")
	if err != nil {
		return nil, err
	}
	profile := profileDoc.Profile

	catalogDoc, err := ParseOscalDocument(catalogInput, "catalog", "oscal-profile-catalog")
	if err != nil {
		return nil, err
	}
	catalog := catalogDoc.Catalog

	// Validate: single import only
	if len(profile.Imports) == 0 {
		return nil, fmt.Errorf("oscal-profile: profile has no imports")
	}
	if len(profile.Imports) > 1 {
		return nil, fmt.Errorf("oscal-profile: profile has %d imports — this converter only supports single-catalog imports. Use NIST's oscal-cli to resolve complex profiles, or use a pre-resolved catalog", len(profile.Imports))
	}

	// Collect included control IDs
	imp := profile.Imports[0]
	includedIDs := collectIncludedIDs(imp)

	// Collect excluded control IDs
	excludedIDs := collectExcludedIDs(imp)

	// Filter catalog
	resolvedCatalog := filterCatalog(catalog, includedIDs, excludedIDs)

	// Apply alter directives (add/remove parts and props on controls)
	if profile.Modify != nil && len(profile.Modify.Alters) > 0 {
		applyAlters(resolvedCatalog, profile.Modify.Alters)
	}

	// Apply parameter overrides from the profile's modify section
	if profile.Modify != nil {
		applyParameterOverrides(resolvedCatalog, profile.Modify.SetParameters)
	}

	// Resolve any remaining {{ insert: param }} placeholders using the
	// catalog's own parameter definitions (select choices, labels).
	resolveDefaultParams(resolvedCatalog)

	// Override metadata from profile (the resolved catalog should carry
	// the profile's metadata, not the source catalog's)
	resolvedCatalog.Metadata = profile.Metadata
	resolvedCatalog.UUID = profile.UUID

	// Compute checksum from the profile input (the profile is the authoritative
	// document; the catalog is a dependency)
	integrity := shared.InputIntegrity(profileInput)

	baseline, err := catalogToBaseline(resolvedCatalog, profileInput, converterVersion)
	if err != nil {
		return nil, fmt.Errorf("oscal-profile: %w", err)
	}

	// Override the integrity to be based on the profile input
	baseline.Integrity = integrity

	return baseline, nil
}

// collectIncludedIDs extracts all control IDs from an import's include-controls.
// Returns nil if no include-controls are specified (meaning include all).
func collectIncludedIDs(imp Import) map[string]bool {
	if len(imp.IncludeControls) == 0 {
		return nil // include all
	}

	ids := make(map[string]bool)
	for _, ic := range imp.IncludeControls {
		for _, id := range ic.WithIDs {
			ids[id] = true
		}
	}
	return ids
}

// collectExcludedIDs extracts all control IDs from an import's exclude-controls.
func collectExcludedIDs(imp Import) map[string]bool {
	if len(imp.ExcludeControls) == 0 {
		return nil
	}

	ids := make(map[string]bool)
	for _, ec := range imp.ExcludeControls {
		for _, id := range ec.WithIDs {
			ids[id] = true
		}
	}
	return ids
}

// filterCatalog creates a new catalog containing only the controls
// matching the include/exclude filters. If includedIDs is nil, all
// controls are included (minus any excludedIDs).
func filterCatalog(catalog *Catalog, includedIDs, excludedIDs map[string]bool) *Catalog {
	includeAll := includedIDs == nil

	result := &Catalog{
		UUID:       catalog.UUID,
		Metadata:   catalog.Metadata,
		BackMatter: catalog.BackMatter,
	}

	for _, group := range catalog.Groups {
		filtered := filterGroupControls(group, includedIDs, excludedIDs, includeAll)
		if len(filtered.Controls) > 0 {
			result.Groups = append(result.Groups, filtered)
		}
	}

	for _, ctrl := range catalog.Controls {
		if shouldIncludeControl(ctrl.ID, includedIDs, excludedIDs, includeAll) {
			filteredCtrl := filterControlEnhancements(ctrl, includedIDs, excludedIDs, includeAll)
			result.Controls = append(result.Controls, filteredCtrl)
		}
	}

	return result
}

// filterGroupControls filters controls within a group.
func filterGroupControls(group Group, includedIDs, excludedIDs map[string]bool, includeAll bool) Group {
	filtered := Group{
		ID:    group.ID,
		Class: group.Class,
		Title: group.Title,
		Props: group.Props,
		Parts: group.Parts,
	}

	for _, ctrl := range group.Controls {
		if shouldIncludeControl(ctrl.ID, includedIDs, excludedIDs, includeAll) {
			filteredCtrl := filterControlEnhancements(ctrl, includedIDs, excludedIDs, includeAll)
			filtered.Controls = append(filtered.Controls, filteredCtrl)
		}
	}

	return filtered
}

// filterControlEnhancements filters nested control enhancements.
func filterControlEnhancements(ctrl Control, includedIDs, excludedIDs map[string]bool, includeAll bool) Control {
	result := Control{
		ID:     ctrl.ID,
		Class:  ctrl.Class,
		Title:  ctrl.Title,
		Params: ctrl.Params,
		Props:  ctrl.Props,
		Links:  ctrl.Links,
		Parts:  ctrl.Parts,
	}

	for _, enh := range ctrl.Controls {
		if shouldIncludeControl(enh.ID, includedIDs, excludedIDs, includeAll) {
			result.Controls = append(result.Controls, enh)
		}
	}

	return result
}

// shouldIncludeControl determines if a control should be included based on
// include/exclude filters.
func shouldIncludeControl(id string, includedIDs, excludedIDs map[string]bool, includeAll bool) bool {
	if excludedIDs != nil && excludedIDs[id] {
		return false
	}
	if includeAll {
		return true
	}
	return includedIDs[id]
}

// applyParameterOverrides applies set-parameters from the profile's modify
// section to the resolved catalog's controls.
func applyParameterOverrides(catalog *Catalog, setParams []SetParameter) {
	if len(setParams) == 0 {
		return
	}

	// Build a param-id → values lookup
	overrides := make(map[string][]string, len(setParams))
	for _, sp := range setParams {
		overrides[sp.ParamID] = sp.Values
	}

	// Apply to all controls in groups
	for i := range catalog.Groups {
		for j := range catalog.Groups[i].Controls {
			applyParamOverridesToControl(&catalog.Groups[i].Controls[j], overrides)
			for k := range catalog.Groups[i].Controls[j].Controls {
				applyParamOverridesToControl(&catalog.Groups[i].Controls[j].Controls[k], overrides)
			}
		}
	}

	// Apply to top-level controls
	for i := range catalog.Controls {
		applyParamOverridesToControl(&catalog.Controls[i], overrides)
		for j := range catalog.Controls[i].Controls {
			applyParamOverridesToControl(&catalog.Controls[i].Controls[j], overrides)
		}
	}
}

// applyParamOverridesToControl replaces parameter prose with override values
// in the control's parts. Parameter references in OSCAL prose look like
// {{ insert: param, param-id }}. We replace these with the override values.
func applyParamOverridesToControl(ctrl *Control, overrides map[string][]string) {
	for i, param := range ctrl.Params {
		if values, ok := overrides[param.ID]; ok {
			// Update the param's label/guidelines to reflect the override
			ctrl.Params[i].Label = strings.Join(values, ", ")
		}
	}

	// Replace parameter insertions in prose
	for i := range ctrl.Parts {
		replaceParamInsertions(&ctrl.Parts[i], overrides)
	}
}

// replaceParamInsertions replaces {{ insert: param, <id> }} patterns in prose
// with the override values.
func replaceParamInsertions(part *Part, overrides map[string][]string) {
	if part.Prose != "" {
		part.Prose = substituteParams(part.Prose, overrides)
	}
	for i := range part.Parts {
		replaceParamInsertions(&part.Parts[i], overrides)
	}
}

// substituteParams replaces OSCAL parameter insertion patterns in text.
// Pattern: {{ insert: param, <param-id> }}
func substituteParams(text string, overrides map[string][]string) string {
	result := text
	for paramID, values := range overrides {
		// OSCAL uses {{ insert: param, param-id }} syntax
		placeholder := fmt.Sprintf("{{ insert: param, %s }}", paramID)
		replacement := strings.Join(values, ", ")
		result = strings.ReplaceAll(result, placeholder, replacement)
	}
	return result
}

// applyAlters applies alter directives to a resolved catalog.
// Each alter targets a control by ID and can add or remove parts and props.
func applyAlters(catalog *Catalog, alters []Alter) {
	for _, alter := range alters {
		ctrl := findControl(catalog, alter.ControlID)
		if ctrl == nil {
			continue // control not in resolved catalog (filtered out)
		}

		// Apply removes first (OSCAL processing order)
		for _, remove := range alter.Removes {
			if remove.ByID != "" {
				ctrl.Parts = removePartByID(ctrl.Parts, remove.ByID)
			}
			if remove.ByName != "" {
				ctrl.Props = removePropByName(ctrl.Props, remove.ByName)
			}
		}

		// Apply adds
		for _, add := range alter.Adds {
			if len(add.Props) > 0 {
				ctrl.Props = append(ctrl.Props, add.Props...)
			}
			if len(add.Parts) > 0 {
				if add.ByID != "" {
					// Add to a specific part identified by by-id
					addPartsToTarget(&ctrl.Parts, add.ByID, add.Parts, add.Position)
				} else {
					// Add directly to the control's parts
					ctrl.Parts = addPartsAtPosition(ctrl.Parts, add.Parts, add.Position)
				}
			}
		}
	}
}

// findControl locates a control by ID in the catalog (groups + top-level).
func findControl(catalog *Catalog, id string) *Control {
	for i := range catalog.Groups {
		for j := range catalog.Groups[i].Controls {
			if catalog.Groups[i].Controls[j].ID == id {
				return &catalog.Groups[i].Controls[j]
			}
			// Check enhancements
			for k := range catalog.Groups[i].Controls[j].Controls {
				if catalog.Groups[i].Controls[j].Controls[k].ID == id {
					return &catalog.Groups[i].Controls[j].Controls[k]
				}
			}
		}
	}
	for i := range catalog.Controls {
		if catalog.Controls[i].ID == id {
			return &catalog.Controls[i]
		}
		for j := range catalog.Controls[i].Controls {
			if catalog.Controls[i].Controls[j].ID == id {
				return &catalog.Controls[i].Controls[j]
			}
		}
	}
	return nil
}

// removePartByID removes a part (and its children) by ID from a parts tree.
func removePartByID(parts []Part, id string) []Part {
	var result []Part
	for _, p := range parts {
		if p.ID == id {
			continue // remove this part
		}
		p.Parts = removePartByID(p.Parts, id) // recurse into children
		result = append(result, p)
	}
	return result
}

// removePropByName removes all props matching a name.
func removePropByName(props []Property, name string) []Property {
	var result []Property
	for _, p := range props {
		if p.Name != name {
			result = append(result, p)
		}
	}
	return result
}

// addPartsToTarget finds a part by ID in the tree and adds child parts to it.
func addPartsToTarget(parts *[]Part, targetID string, newParts []Part, position string) {
	for i := range *parts {
		if (*parts)[i].ID == targetID {
			(*parts)[i].Parts = addPartsAtPosition((*parts)[i].Parts, newParts, position)
			return
		}
		// Recurse
		addPartsToTarget(&(*parts)[i].Parts, targetID, newParts, position)
	}
}

// resolveDefaultParams substitutes remaining {{ insert: param, X }} placeholders
// using the catalog's own parameter definitions. For params with select choices,
// produces "[Selection: <how-many>: choice1; choice2]". For params with only a
// label, produces "[Assignment: <label>]".
func resolveDefaultParams(catalog *Catalog) {
	defaults := collectParamDefaults(catalog)
	if len(defaults) == 0 {
		return
	}
	// Apply defaults to all controls (same walk as applyParameterOverrides)
	for i := range catalog.Groups {
		for j := range catalog.Groups[i].Controls {
			applyParamOverridesToControl(&catalog.Groups[i].Controls[j], defaults)
			for k := range catalog.Groups[i].Controls[j].Controls {
				applyParamOverridesToControl(&catalog.Groups[i].Controls[j].Controls[k], defaults)
			}
		}
	}
	for i := range catalog.Controls {
		applyParamOverridesToControl(&catalog.Controls[i], defaults)
		for j := range catalog.Controls[i].Controls {
			applyParamOverridesToControl(&catalog.Controls[i].Controls[j], defaults)
		}
	}
}

// collectParamDefaults builds a param-id → fallback-values map from the catalog's
// own parameter definitions. Profile set-parameters take precedence (applied first).
func collectParamDefaults(catalog *Catalog) map[string][]string {
	defaults := make(map[string][]string)
	var collectFromControls func(controls []Control)
	collectFromControls = func(controls []Control) {
		for _, ctrl := range controls {
			for _, param := range ctrl.Params {
				if param.Select != nil && len(param.Select.Choice) > 0 {
					howMany := param.Select.HowMany
					if howMany == "" {
						howMany = "one or more"
					}
					fallback := fmt.Sprintf("[Selection (%s): %s]", howMany, strings.Join(param.Select.Choice, "; "))
					defaults[param.ID] = []string{fallback}
				} else if param.Label != "" {
					defaults[param.ID] = []string{fmt.Sprintf("[Assignment: %s]", param.Label)}
				}
			}
			collectFromControls(ctrl.Controls)
		}
	}
	for _, group := range catalog.Groups {
		collectFromControls(group.Controls)
	}
	collectFromControls(catalog.Controls)
	return defaults
}

// addPartsAtPosition inserts parts at the specified position.
func addPartsAtPosition(existing, newParts []Part, position string) []Part {
	switch position {
	case "starting":
		return append(newParts, existing...)
	case "before":
		return append(newParts, existing...)
	case "ending", "":
		return append(existing, newParts...)
	case "after":
		return append(existing, newParts...)
	default:
		return append(existing, newParts...)
	}
}

File: hdf-converters/converters/oscal-to-hdf/typescript/converter-profile.ts
/**
 * OSCAL Profile to HDF Baseline converter.
 *
 * Mirrors the Go implementation in converters/oscal-to-hdf/go/converter_profile.go.
 *
 * This implements the "simple resolver" -- it handles:
 * - A single catalog import
 * - include-controls with with-ids filtering
 * - set-parameters from modify section
 * - merge as-is (preserve catalog structure)
 *
 * It does NOT handle:
 * - Multiple imports
 * - Nested profile imports
 * - alter directives
 * - Complex merge strategies
 */

import { parseJSON } from '@mitre/hdf-utilities';
import { inputIntegrity, validateInputSize } from '../../../shared/typescript/converterutil.js';
import type {
  Oscal,
  Catalog,
  Control,
  CatalogControlGroup,
  ImportResource,
  Part,
  ParameterSetting,
} from './types.js';
import { catalogToBaseline } from './converter-catalog.js';

/**
 * Converts an OSCAL Profile against a provided catalog to HDF Baseline JSON.
 *
 * @param profileInput - Raw JSON string containing an OSCAL profile
 * @param catalogInput - Raw JSON string containing an OSCAL catalog
 * @returns HDF Baseline JSON string
 */
export async function convertOscalProfileToHdf(
  profileInput: string,
  catalogInput: string,
): Promise<string> {
  validateInputSize(profileInput, 'oscal-profile');

  if (!profileInput || profileInput.trim().length === 0) {
    throw new Error('empty profile input');
  }
  if (!catalogInput || catalogInput.trim().length === 0) {
    throw new Error('empty catalog input');
  }

  // Parse profile
  const profileDoc = parseJSON<Oscal>(profileInput);
  if (!profileDoc.profile) {
    throw new Error(
      "oscal-profile: input is not a profile document (root key is not 'profile')",
    );
  }
  const profile = profileDoc.profile;

  // Parse catalog
  const catalogDoc = parseJSON<Oscal>(catalogInput);
  if (!catalogDoc.catalog) {
    throw new Error(
      "oscal-profile: catalog input is not a catalog document (root key is not 'catalog')",
    );
  }
  const catalog = catalogDoc.catalog;

  // Validate: single import only
  if (!profile.imports || profile.imports.length === 0) {
    throw new Error('oscal-profile: profile has no imports');
  }
  if (profile.imports.length > 1) {
    throw new Error(
      `oscal-profile: profile has ${profile.imports.length} imports -- this converter only supports single-catalog imports. Use NIST's oscal-cli to resolve complex profiles, or use a pre-resolved catalog`,
    );
  }

  // Validate: no alter directives
  if (profile.modify?.alters && profile.modify.alters.length > 0) {
    throw new Error(
      `oscal-profile: profile contains ${profile.modify.alters.length} alter directives -- this converter only supports parameter overrides. Use NIST's oscal-cli to resolve profiles with alter directives, or use a pre-resolved catalog`,
    );
  }

  // Collect included/excluded control IDs
  const imp = profile.imports[0]!;
  const includedIDs = collectIncludedIDs(imp);
  const excludedIDs = collectExcludedIDs(imp);

  // Filter catalog
  const resolvedCatalog = filterCatalog(catalog, includedIDs, excludedIDs);

  // Apply parameter overrides
  if (profile.modify?.['set-parameters']) {
    applyParameterOverrides(resolvedCatalog, profile.modify['set-parameters']);
  }

  // Override metadata from profile
  resolvedCatalog.metadata = profile.metadata;
  resolvedCatalog.uuid = profile.uuid;

  const baseline = await catalogToBaseline(resolvedCatalog, profileInput);

  // Override the integrity to be based on the profile input
  baseline.integrity = await inputIntegrity(profileInput);

  return JSON.stringify(baseline, null, 2);
}

/** Extracts all control IDs from an import's include-controls. Returns null if include all. */
function collectIncludedIDs(imp: ImportResource): Set<string> | null {
  const includeControls = imp['include-controls'];
  if (!includeControls || includeControls.length === 0) {
    return null; // include all
  }

  const ids = new Set<string>();
  for (const ic of includeControls) {
    for (const id of ic['with-ids'] ?? []) {
      ids.add(id);
    }
  }
  return ids;
}

/** Extracts all control IDs from an import's exclude-controls. */
function collectExcludedIDs(imp: ImportResource): Set<string> | null {
  const excludeControls = imp['exclude-controls'];
  if (!excludeControls || excludeControls.length === 0) {
    return null;
  }

  const ids = new Set<string>();
  for (const ec of excludeControls) {
    for (const id of ec['with-ids'] ?? []) {
      ids.add(id);
    }
  }
  return ids;
}

/** Creates a new catalog containing only the controls matching filters. */
function filterCatalog(
  catalog: Catalog,
  includedIDs: Set<string> | null,
  excludedIDs: Set<string> | null,
): Catalog {
  const includeAll = includedIDs === null;

  const result: Catalog = {
    uuid: catalog.uuid,
    metadata: catalog.metadata,
    'back-matter': catalog['back-matter'],
    groups: [],
    controls: [],
  };

  for (const group of catalog.groups ?? []) {
    const filtered = filterGroupControls(group, includedIDs, excludedIDs, includeAll);
    if (filtered.controls && filtered.controls.length > 0) {
      result.groups!.push(filtered);
    }
  }

  for (const ctrl of catalog.controls ?? []) {
    if (shouldIncludeControl(ctrl.id, includedIDs, excludedIDs, includeAll)) {
      const filteredCtrl = filterControlEnhancements(ctrl, includedIDs, excludedIDs, includeAll);
      result.controls!.push(filteredCtrl);
    }
  }

  return result;
}

function filterGroupControls(
  group: CatalogControlGroup,
  includedIDs: Set<string> | null,
  excludedIDs: Set<string> | null,
  includeAll: boolean,
): CatalogControlGroup {
  const filtered: CatalogControlGroup = {
    id: group.id,
    class: group.class,
    title: group.title,
    props: group.props,
    parts: group.parts,
    controls: [],
  };

  for (const ctrl of group.controls ?? []) {
    if (shouldIncludeControl(ctrl.id, includedIDs, excludedIDs, includeAll)) {
      const filteredCtrl = filterControlEnhancements(ctrl, includedIDs, excludedIDs, includeAll);
      filtered.controls!.push(filteredCtrl);
    }
  }

  return filtered;
}

function filterControlEnhancements(
  ctrl: Control,
  includedIDs: Set<string> | null,
  excludedIDs: Set<string> | null,
  includeAll: boolean,
): Control {
  const result: Control = {
    id: ctrl.id,
    class: ctrl.class,
    title: ctrl.title,
    params: ctrl.params,
    props: ctrl.props,
    links: ctrl.links,
    parts: ctrl.parts,
    controls: [],
  };

  for (const enh of ctrl.controls ?? []) {
    if (shouldIncludeControl(enh.id, includedIDs, excludedIDs, includeAll)) {
      result.controls!.push(enh);
    }
  }

  return result;
}

function shouldIncludeControl(
  id: string,
  includedIDs: Set<string> | null,
  excludedIDs: Set<string> | null,
  includeAll: boolean,
): boolean {
  if (excludedIDs !== null && excludedIDs.has(id)) {
    return false;
  }
  if (includeAll) {
    return true;
  }
  return includedIDs !== null && includedIDs.has(id);
}

/** Applies set-parameters from the profile's modify section to the resolved catalog. */
function applyParameterOverrides(
  catalog: Catalog,
  setParams: ParameterSetting[],
): void {
  if (setParams.length === 0) return;

  // Build param-id -> values lookup
  const overrides = new Map<string, string[]>();
  for (const sp of setParams) {
    if (sp['param-id']) {
      overrides.set(sp['param-id'], sp.values ?? []);
    }
  }

  // Apply to all controls in groups
  for (const group of catalog.groups ?? []) {
    for (const ctrl of group.controls ?? []) {
      applyParamOverridesToControl(ctrl, overrides);
      for (const enh of ctrl.controls ?? []) {
        applyParamOverridesToControl(enh, overrides);
      }
    }
  }

  // Apply to top-level controls
  for (const ctrl of catalog.controls ?? []) {
    applyParamOverridesToControl(ctrl, overrides);
    for (const enh of ctrl.controls ?? []) {
      applyParamOverridesToControl(enh, overrides);
    }
  }
}

function applyParamOverridesToControl(
  ctrl: Control,
  overrides: Map<string, string[]>,
): void {
  if (ctrl.params) {
    for (const param of ctrl.params) {
      if (!param.id) continue;
      const values = overrides.get(param.id);
      if (values) {
        param.label = values.join(', ');
      }
    }
  }

  // Replace parameter insertions in prose
  if (ctrl.parts) {
    for (const part of ctrl.parts) {
      replaceParamInsertions(part, overrides);
    }
  }
}

function replaceParamInsertions(part: Part, overrides: Map<string, string[]>): void {
  if (part.prose) {
    part.prose = substituteParams(part.prose, overrides);
  }
  if (part.parts) {
    for (const child of part.parts) {
      replaceParamInsertions(child, overrides);
    }
  }
}

/** Replaces OSCAL parameter insertion patterns: {{ insert: param, <param-id> }} */
function substituteParams(text: string, overrides: Map<string, string[]>): string {
  let result = text;
  for (const [paramId, values] of overrides) {
    const placeholder = `{{ insert: param, ${paramId} }}`;
    const replacement = values.join(', ');
    result = result.split(placeholder).join(replacement);
  }
  return result;
}
